docs/develop/typescript/message-passing.mdx

```diff
@@ -196,107 +290,319 @@ await client.workflow.signalWithStart(yourWorkflow, {
 
-async function run(): Promise<void> {
-  const client = new Client();
-  const handle = client.workflow.getHandle('state-id-0');
-  const meaning = await handle.query(getValueQuery, 'meaning-of-life');
-  console.log({ meaning });
+#### Ensure your handlers finish before the Workflow completes {#wait-for-message-handlers}
+
+Workflow wait conditions can ensure your handler completes before a Workflow finishes.
+When your Workflow uses `async` Signal or Update handlers, your main Workflow method can return or Continue-as-New while a handler is still waiting on an async task, such as an Activity result.
+The Workflow completing may interrupt the handler before it finishes crucial work and cause client errors when trying retrieve Update results.
+Use [`workflow.condition`](https://typescript.temporal.io/api/namespaces/workflow#condition) and [`allHandlersFinished`](https://typescript.temporal.io/api/namespaces/workflow#condition#allhandlersfinished) to address this problem and allow your Workflow to end smoothly:
+
+```typescript
+export async function myWorkflow(): Promise<MyWorkflowOutput> {
+  await wf.condition(wf.allHandlersFinished);
+  return workflowOutput;
 }
 ```
 
-<!--SNIPEND-->
+By default, your Worker will log a warning when you allow a Workflow Execution to finish with unfinished handler executions.
+You can silence these warnings on a per-handler basis by setting the `unfinishedPolicy` in `SignalHandlerOptions` or `UpdateHandlerOptions` when calling [`workflow.setHandler`](https://typescript.temporal.io/api/namespaces/workflow#sethandler)
+
+See [Finishing handlers before the Workflow completes](/encyclopedia/workflow-message-passing#finishing-message-handlers) for more information.
+
+### Use `asyncio.Lock` to prevent concurrent handler execution {#control-handler-concurrency}
+
+Concurrent processes can interact in unpredictable ways.
+Incorrectly written [concurrent message-passing](/encyclopedia/workflow-message-passing#message-handler-concurrency) code may not work correctly when multiple handler instances run simultaneously.
+Here's an example of a pathological case:
+
+```typescript
+export async function myWorkflow(): Promise<MyWorkflowOutput> {
+  let x = 0;
+  let y = 0;
+  wf.setHandler(mySignal, async () => {
+    const data = await myActivity();
+    x = data.x;
+
+    // 🐛🐛 Bug!! If multiple instances of this handler are executing
+    // concurrently, then there may be times when the Workflow has x from one
+    // Activity execution and y from another.
+    await wf.sleep(500); // or await anything else
+
+    y = data.y;
+  });
+  ...
+}
+```
+
+Coordinating access using a lock (also known as a mutex) corrects this code.
+Locking makes sure that only one handler instance can execute a specific section of code at any given time:
+
+```typescript
+export async function myWorkflow(): Promise<MyWorkflowOutput> {
+  let x = 0;
+  let y = 0;
+  const lock = new Mutex();
```

Comment:
Please add an import statement for `Mutex`, and should even link to the NPM package.

Reply:
Added. I guess there's nothing we can do about the author's GitHub handle.